KAFKA-17094: Return fenced brokers in DescribeCluster response

tinaselenge opened the pull request:
Implementation of KIP-1073. 

Add new unit and integration tests for describeCluster.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

tinaselenge committed:
KIP-1073: Return fenced brokers in DescribeCluster response

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/ClusterTool.java:
```diff
@@ -108,6 +117,17 @@ static void execute(String... args) throws Exception {
+                    boolean includeFencedBrokers = Optional.of(namespace.getBoolean("include_fenced_brokers")).orElse(false);
+                    boolean listControllerEndpoints = namespace.getString("bootstrap_controller") != null;
+                    if (includeFencedBrokers && listControllerEndpoints) {
+                        throw new RuntimeException("Cannot use --include-fenced-brokers with command " + command);
```
```suggestion
                        throw new IllegalArgumentException("The option --include-fenced-brokers is only supported with --bootstrap-server option");
```

fvaleri reviewed (commented):
Hi @tinaselenge, thanks for working on this useful feature.

There are few minor style issues in ClusterTool, and we need a rebase.

----

I created a KRaft cluster with 3 brokers (7, 8, 9) and dynamic quorum. Then, I stopped broker 8, which is fenced as expected:

```sh
| offset: 629 CreateTime: 1730803001463 keySize: -1 valueSize: 19 sequence: -1 headerKeys: [] payload: {"type":"BROKER_REGISTRATION_CHANGE_RECORD","version":0,"data":{"brokerId":8,"brokerEpoch":26,"fenced":1}}
```

After some time, it disappears from describe quorum. This is also expected.

```sh
NodeId	DirectoryId           	LogEndOffset	Lag	LastFetchTimestamp	LastCaughtUpTimestamp	Status  	
0     	knuNd8k4S9u-8YMnEOdofw	1595        	0  	4 ms ago          	5 ms ago             	Leader  	
1     	dnAw7IxCTI-vxgYU-NiYDQ	1595        	0  	386 ms ago        	386 ms ago           	Follower	
2     	C_a66WN8S_mT4b4l6_Ca3Q	1595        	0  	386 ms ago        	386 ms ago           	Follower	
9     	_yapG4z_YyfcdhDhWR3B9w	1595        	0  	386 ms ago        	386 ms ago           	Observer	
7     	ipPrA6ueF48LcwEDWXYNYA	1595        	0  	386 ms ago        	386 ms ago           	Observer
```

At this point, if I run `list-endpoints` without `--include-fenced-brokers` flag the output looks good.

```sh
ID         HOST      PORT       RACK       STATE      ENDPOINT_TYPE  
7          localhost 9092       null       unfenced   broker         
9          localhost 9094       null       unfenced   broker 
```

Instead, this is the output when I use the `--include-fenced-brokers` flag. Broker 8 is there, which is what we want, but the state is unfenced, not fenced as I would expect.

```sh
ID         HOST      PORT       RACK       STATE      ENDPOINT_TYPE  
7          localhost 9092       null       unfenced   broker         
8          localhost 9093       null       unfenced   broker         
9          localhost 9094       null       unfenced   broker    
```

tinaselenge committed:
Fix the issue reported in the review comment

Rebase to trunk and resolve conflicts
Refine the tests to make sure that brokers have the expected fenced state
Add tests for listing nodes from controller (this needed to be in SSLAdminIntegrationTest because of how useBoostrapControllers is set up)

tinaselenge commented:
@fvaleri thank you so much for spotting that! This was because my change for DescribeClusterResponse.class was missed out from the commit accidentally. This issue also pointed out that the tests needed to be improved, which I did now.

fvaleri commented on core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:
```diff
@@ -615,6 +615,42 @@ class PlaintextAdminIntegrationTest extends BaseAdminIntegrationTest {
+    client = createAdminClient
+    val fencedBrokerId = brokers.last.config.brokerId
+    killBroker(fencedBrokerId)
+    Thread.sleep(10000) //sleep is needed to ensure the broker is fenced after the shutdown
```
Is there any way to avoid this? My fear is that it can create a flaky test when you have a slow environment, like when you run along with a bunch of other tests on the pipeline.

tinaselenge replied:
This is my concern as well. It seems to take at least a few seconds for the broker to become fenced after killing it and was hoping that 10000ms would be more than enough (need to rerun it a number of times in Jenkins to confirm). I understand there is a risk of it not being enough resulting in flakiness. I see that some of the integration tests also use TestUtils.retry() to retry when there is an assertion error, until the max timeout is reached. This still does not guarantee, but I wonder if that's a better solution with a larger max timeout, than the sleep.

tinaselenge replied:
This was my concern too but the broker takes a few seconds to get fenced after being killed and 10000ms seems to be sufficient, although I had planned to run this test several times in Jenkins to confirm it and adjust if needed. 

I pushed another commit to merge these two test cases together, so that we only wait for a broker to get fenced once instead of twice. Also used TestUtils.retry to retry the request until we receive 2 of 3 brokers in the response or reach the max wait time. I set the max wait time to a larger value, in case it takes longer sometimes, but it shouldn't take more than 10s normally.

fvaleri commented on core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:
```diff
@@ -615,6 +615,42 @@ class PlaintextAdminIntegrationTest extends BaseAdminIntegrationTest {
+    client = createAdminClient
+    val fencedBrokerId = brokers.last.config.brokerId
+    killBroker(fencedBrokerId)
+    Thread.sleep(10000) //sleep is needed to ensure the broker is fenced after the shutdown
```
See the other similar comment.

fvaleri reviewed (commented):
Latest changes LGTM, thanks!

I just left a comment in one of the tests. I'll look further into it tomorrow.

tinaselenge committed:
Merge the integration test cases for includeFencedBrokers into one so that we don't have to wait for the broker to get fenced twice and use the TestUtils.retry() for it

fvaleri commented on core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:
```diff
@@ -615,6 +615,30 @@ class PlaintextAdminIntegrationTest extends BaseAdminIntegrationTest {
+    killBroker(fencedBrokerId)
+    // It takes a few seconds for a broker to get fenced after being killed
+    // So we retry until only 2 of 3 brokers returned in the result or the max wait is reached
+    TestUtils.retry(20000) {
```
I think this is ok, because the fencing time should be around `broker.session.timeout.ms` (9 seconds) in the worst case.

https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/controller/QuorumController.java#L1650-L1662

One thing we can improve is the to kill the broker immediately:

```sh
killBroker(fencedBrokerId, JDuration.ofMillis(0))
```

fvaleri reviewed (commented):
LGTM, but there are still some minor check style issues that you can easily identify by running: 

```sh
./gradlew tools:checkstyleMain ; ./gradlew tools:spotlessCheck
```

tinaselenge committed:
Fix checkstyle errors

fvaleri reviewed (approved):
LGTM. Thanks.

tinaselenge commented:
thanks a lot @fvaleri!

showuon commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -3611,12 +3611,17 @@ class KafkaApis(val requestChannel: RequestChannel,
+        val describeClusterRequest = request.body[DescribeClusterRequest]
+        metadataCache.getBrokerNodes(request.context.listenerName).foreach { node =>
+          if (node.isFenced && !describeClusterRequest.data().includeFencedBrokers()) {
+          } else {
```
What do we want to do in the if block? log something? Or remove if / else?

tinaselenge replied:
I removed the else and swapped the condition around

showuon commented on core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala:
```diff
@@ -354,6 +354,10 @@ class ZkMetadataCache(
     metadataSnapshot.aliveBrokers.values.flatMap(_.getNode(listenerName))
   }
 
+  override def getBrokerNodes(listenerName: ListenerName): Iterable[Node] = {
+    getAliveBrokerNodes(listenerName)
+  }
+
```
Since we're targeting v4.0.0, this can be removed.

tinaselenge replied:
It seems like until we remove ZKMetadataCache completely, this is needed in order to compile.

showuon replied:
Fair enough.

showuon commented on tools/src/test/java/org/apache/kafka/tools/ClusterToolTest.java:
```diff
@@ -69,6 +69,28 @@ public void testUnregister(ClusterInstance clusterInstance) {
+    }
+
+    @ClusterTest(brokers = 1, types = {Type.KRAFT, Type.CO_KRAFT})
+    public void testListEndpointsArgumentWithBootstrapServer(ClusterInstance clusterInstance) {
```
Could we add a test to make sure the "fenced" broker state can be correctly returned and output?

showuon commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -3194,6 +3194,25 @@ public void testListGroupsWithTypesOlderBrokerVersion() throws Exception {
+            final DescribeClusterResult result = env.adminClient().describeCluster(new DescribeClusterOptions().includeFencedBrokers(true));
+            TestUtils.assertFutureThrows(result.nodes(),
+                    UnsupportedVersionException.class, "Including fenced broker endpoints is not supported with version 1");
+
```
nit: additional empty line.

showuon commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeClusterRequest.java:
```diff
@@ -51,6 +52,13 @@ public String toString() {
     public DescribeClusterRequest(DescribeClusterRequestData data, short version) {
         super(ApiKeys.DESCRIBE_CLUSTER, version);
         this.data = data;
+        validateVersionForIncludingFencedBrokers(version, data.includeFencedBrokers());
+    }
+
+    private void validateVersionForIncludingFencedBrokers(short version, boolean includeFencedBrokers) {
+        if (version < 2 && includeFencedBrokers) {
+            throw new UnsupportedVersionException("Including fenced broker endpoints is not supported with version " + version);
+        }
```
In the generated clients/build/generated/main/java/org/apache/kafka/common/message/DescribeClusterRequestData.java, the `UnsupportedVersionException` will be thrown when trying to add `includeFencedBrokers` into version < 2, so we don't have to do it by our own here.

tinaselenge replied:
This doesn't seem to be the case when I tested it by running new AdminClient against an old broker. For example, if we remove this validation, testDescribeClusterHandleUnsupportedVersionForIncludingFencedBrokers() test fails. Also this helps Admin to handle the UnsupportedVersion exception differently, otherwise we fall back to metadata request and it's not clear to users why the request the failed.

showuon replied:
1. Currently, the `testDescribeClusterHandleUnsupportedVersionForIncludingFencedBrokers` will return `UnsupportedVersionException` with `Api DESCRIBE_CLUSTER with version 1` from [here](https://github.com/apache/kafka/blob/48d60efe9802786451458aec9136b7b585fecd92/clients/src/test/java/org/apache/kafka/clients/MockClient.java#L261). That's why it will fail. So, basically, it will fail with `UnsupportedVersionException` as what we expected. 

2. I understand why we need this different message as described in this [comment](https://github.com/apache/kafka/pull/17524#discussion_r1850837344).
> We fallback to metadata request, if the broker does not support the new DescribeCluster API. In this case, DescribeCluster API is supported but the option to include fenced broker is not. So this is why we are checking the specific error message to differentiate the cause of UnsupportedVersion exception.

But I think using the error message as a condition is really weak. I think, like you said, the reason we can't fallback to metadata request is because the user added an option to _include fenced broker_. That is, when user requests to "include fenced broker", this is the case that we should not fallback to metadata request, even if it returned `UnsupportedVersionException`. After all, the metadata request response won't include fenced broker, it is meaningless to users. So, like how we treat `usingBootstrapControllers`, we can do similar thing to this case, WDYT?

That is:
```
boolean handleUnsupportedVersionException(final UnsupportedVersionException exception) {
  if (metadataManager.usingBootstrapControllers()) {
     return false;
  }
  // add some comments here
  if (options.includeFencedBrokers()) {
    return false;
  }
```

Also, we should have an integration test for this unsupported version case. Like [this test](https://github.com/apache/kafka/blob/48d60efe9802786451458aec9136b7b585fecd92/core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala#L1354) did.

tinaselenge replied:
I agree that checking the option instead of the error message is much better solution here. I made an update to add an error level log, instead of validating describe request and throwing a specific error for includeFencedBrokers option. 

In terms of the integration test, I'm going to look into to it a little more. KRaftClusterTest.scala you pointed, seems like more cluster level tests. Since this is testing Admin specific option, it should probably be added to AdminIntegrationTest but I'm currently not sure how to set the test cluster's version or it's supported API version.

showuon replied:
Hmm... it's quite complicated to have a integration test to test with lower describeCluster version. I think it's fine sine we already have unit test to cover it. Thanks.

tinaselenge replied:
I agree, I think this will take quite a lot of refactoring to the testing framework. We have unit tests for the compatibility part and integration tests for the new option to include fenced brokers.

showuon commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2554,6 +2557,10 @@ boolean handleUnsupportedVersionException(final UnsupportedVersionException exce
                     return false;
                 }
 
+                if (exception.getMessage().contains("Including fenced broker endpoints is not supported with version")) {
+                    return false;
+                }
+
                 useMetadataRequest = true;
                 return true;
```
So, in [KIP-700](https://cwiki.apache.org/confluence/display/KAFKA/KIP-700%3A+Add+Describe+Cluster+API), we implement a fallback to metadata request mechanism when describe cluster failed with unsupported version. So if we failed due to `Including fenced broker endpoints is not supported with version`, shouldn't we fallback to use metadata request? And the follow-up question, what will it be in the admin client result?

tinaselenge replied:
Thanks Luke for reviewing the PR. I'm not quite sure on this. We fallback to metadata request, if the broker does not support the new DescribeCluster API. In this case, DescribeCluster API is supported but the option to include fenced broker is not. So this is why we are checking the specific error message to differentiate the cause of UnsupportedVersion exception.

showuon reviewed (commented):
Made a pass, left some comments. Thanks.

tinaselenge committed:
Address review comment

tinaselenge committed:
Address review comment

showuon commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2557,7 +2557,10 @@ boolean handleUnsupportedVersionException(final UnsupportedVersionException exce
+                // If unsupportedVersion exception was caused by the option to include fenced brokers (only supported for version 1+)
+                // then we should not fall back to the metadataRequest.
+                if (options.includeFencedBrokers()) {
+                    log.error("Including fenced broker endpoints is not supported with " + exception.getMessage());
```
I'm not sure the exception message when unsupportedVersionException thrown. You need to check again in integration test. The `Api DESCRIBE_CLUSTER with version 1` is from `mockClient`, not real `networkClient`. FYI.

tinaselenge replied:
Right. The actual error message you get is `Attempted to write a non-default includeFencedBrokers at version 1`. Perhaps this is self explanatory so don't need this log?

showuon replied:
I think so.

showuon commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2554,6 +2557,12 @@ boolean handleUnsupportedVersionException(final UnsupportedVersionException exce
                     return false;
                 }
 
+                // If unsupportedVersion exception was caused by the option to include fenced brokers (only supported for version 1+)
```
supported for version 2+?

showuon reviewed (approved):
LGTM! Thanks for the PR!

tinaselenge commented:
thank you @showuon!

showuon commented:
@tinaselenge , there are some tests failure. Could you help fix them? Thanks.

tinaselenge committed:
fix unit test

tinaselenge commented:
Thanks @showuon , I pushed a fix for the failed test.
